### PR TITLE
Pass membership data in form_state for roles add/remove form.

### PR DESCRIPTION
**Pull request by dipakmdhrm**
#749

### CODE REVIEW

**Comment by amitaibu** [comment]
I think a better solution is what you proposed in https://github.com/Gizra/og/issues/749

> Update the current paths for role add & remove from /group/roles/add & /group/roles/remove to /group/{group}/roles/add/{member} & /group/{group}/roles/remove/{member} respectively. This will allow developers to get the group & member data from path.

**Comment by dipakmdhrm** [comment]
> I think a better solution is what you proposed in #749
> 
> > Update the current paths for role add & remove from /group/roles/add & /group/roles/remove to /group/{group}/roles/add/{member} & /group/{group}/roles/remove/{member} respectively. This will allow developers to get the group & member data from path.

Implementing this wouldn't be possible because of use of actions to add/remove roles. 

og-actions

This is a problem for our proposed implementation because of 2 reasons:
1. My proposed implementation (`/group/{group}/roles/add/{member}`) only considers addition/removal of roles for just 1 member at a time. However, actions allows user to select multiple members and bulk update roles for them.
2. Actions (example, publish node, set flag etc.) are supposed to be self contained and are not usually expected to pass data in a next step to some form. Actions are in no way connected to a new form on a new page (like our proposed form at /group/{group}/add/roles) and we need some way to pass the list of selected members to the form. That's why we need tempstore. 

I tried to find other modules where something similar was set up in hope I might find a way. But even in other modules (views_bulk_edit for example) tempstore is used.

That unfortunately leaves us with the fallback of passing membership data from form build to form alter for people to be able to use it.

In conclusion: I ended up spending way too much time only to end up with disappointment :sweat_smile:

**Comment by amitaibu** [comment]
Having it in the form_state then would be fine :+1:

**Comment by amitaibu** [comment]
@dipakmdhrm this needs more work

**Comment by dipakmdhrm** [comment]
Added examples in `og.api.php`

File: og.api.php [hunk 1]
```diff
@@ -54,6 +55,34 @@ function hook_og_user_access_alter(array &$permissions, CacheableMetadata $cache
   $cacheable_metadata->addCacheableDependency($config);
 }
 
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function hook_form_og_membership_remove_multiple_roles_action_form_alter(array &$form, FormStateInterface $form_state, string $form_id) {
+  // Get access to current group and selected memberships when on role remove
```
Comment (amitaibu) [inline og.api.php]:
```suggestion
  // Get access to current group and selected memberships when we're on role remove
```

File: src/Form/OgChangeMultipleRolesFormBase.php [hunk 2]
```diff
@@ -93,6 +93,17 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       '#value' => $this->t('Submit'),
     ];
 
+    // Set selected memberships & group in form state. This can be used by
+    // developers to alter the roles form.
```
Comment (amitaibu) [inline src/Form/OgChangeMultipleRolesFormBase.php]:
```suggestion
    // Set selected memberships in the form state. This can be used by
    // to alter the roles form.
```

File: src/Form/OgChangeMultipleRolesFormBase.php [hunk 3]
```diff
@@ -93,6 +93,17 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
+    $memberships = $this->getMemberships();
+    if (empty($memberships)) {
+      return $form;
+    }
+    $membership = current($memberships);
+    $group = $membership->getGroup();
```
Comment (amitaibu) [inline src/Form/OgChangeMultipleRolesFormBase.php]: Can this be removed? We can extract the group from the Membership

**Review by amitaibu — COMMENTED** [review]
Still WIP?

File: src/Form/OgChangeMultipleRolesFormBase.php (hunk 3 above)
Comment (dipakmdhrm) [reply src/Form/OgChangeMultipleRolesFormBase.php]: Removed

**Comment by amitaibu** [comment]
@dipakmdhrm can you please work on a PR to fix the Travis errors (against the `8.x-1.x` branch)

**Comment by dipakmdhrm** [comment]
> @dipakmdhrm can you please work on a PR to fix the Travis errors (against the `8.x-1.x` branch)

WIP: https://github.com/Gizra/og/pull/763
@amitaibu

**Comment by claudiu-cristea** [comment]
Since https://www.drupal.org/project/og/issues/3467694 was merged into the `8.x-1.x` branch, the development has been moved to https://www.drupal.org/project/og. If this change is still actual, please open an issue at https://www.drupal.org/project/issues/og and provide there a _merge request_ similar with this pull request.

Thank you for contributing to Organic groups module.
